prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts

```diff
@@ -699,4 +656,56 @@ export class EnrolmentFormStateService extends AbstractFormStateService<Enrolmen
       .map((formArrayName: string) => this.selfDeclarationForm.get(formArrayName) as FormArray)
       .forEach((formArray: FormArray) => formArray.clear());
   }
+
+  /**
+   * @description
+   * Check for the requirement of at least one certification, or one obo site/job.
+   */
+  // TODO refactor this method as it can't be scanned and understood
+  private hasCertificateOrOboSite(): boolean {
+    const oboSites = this.oboSitesForm.get('oboSites') as FormArray;
+    const certifications = this.regulatoryFormState.certifications;
+    const enrolleeHealthAuthorities = this.careSettingsForm.get('enrolleeHealthAuthorities') as FormArray;
+    let hasOboSiteForEveryHA = true;
+    // Using `for` loop rather than `every()` method for ease of debugging
+    for (let i = 0; i < enrolleeHealthAuthorities.controls.length; i++) {
+      const checkbox = enrolleeHealthAuthorities.controls[i];
+      // For every selected Health Authority ...
+      if (checkbox.value === true) {
+        let foundMatchingHAOboSite = false;
+        // ... there must be at least one Obo Site for that Health Authority
+        for (let j = 0; j < oboSites.controls.length; j++) {
+          const oboSiteForm = oboSites.controls[j] as FormGroup;
+          if (oboSiteForm.controls.healthAuthorityCode.value === this.configService.healthAuthorities[i].code) {
+            foundMatchingHAOboSite = true;
+            break;
+          }
+        }
+        if (!foundMatchingHAOboSite) {
+          hasOboSiteForEveryHA = false;
+          break;
+        }
+      }
+    }
+    // When you set certifications to 'None' there still exists an item in
+    // the FormArray, and this checks for its existence
+    return (oboSites.length && hasOboSiteForEveryHA) || (certifications.length && certifications.value[0].licenseNumber);
+  }
+
+  /**
+   * @description
+   * Check if the user selected Device Provider as Care Setting
+   * then they shuold provide either Device Provider ID or Obo Site
+   */
+  private hasDeviceProviderOrOboSite(): boolean {
+    const {
+      careSettings,
+      certifications,
+      deviceProviderIdentifier,
+      oboSites } = this.json;
+
+    return (careSettings.some((cs) => cs.careSettingCode === CareSettingEnum.DEVICE_PROVIDER))
+      ? !!(deviceProviderIdentifier && certifications.length || (!deviceProviderIdentifier && oboSites.length))
```

Comment:
Do you feel this reads easily?

```ts
return !!(deviceProviderIdentifier && certifications.length || (!deviceProviderIdentifier && oboSites.length))
```

or is this easier to scan?

```ts
return (deviceProviderIdentifier && !!certifications.length) 
  || (!deviceProviderIdentifier && !!oboSites.length)
```

You have two criteria to check, but when you don't group them and then you invert the whole thing the next person has to unpack it instead of quickly identifying what you're trying to achieve.

Reply:
That was an oversite corrected now